torch/_subclasses/fake_impls.py

```diff
@@ -296,23 +296,59 @@ def unique2(
 
             _constrain_range_for_size(nnz, max=maxval)
 
-        arg.unique_memo = nnz
+        if dim is None:
+            arg.unique_memo = nnz
 
-    ret = [arg.new_empty((nnz,))]
+    if dim is None:
+        ret = [arg.new_empty((nnz,))]
+    else:
+        ret = [arg.new_empty(*arg.shape[:dim], nnz, *arg.shape[dim + 1:])]
 
     if return_inverse:
```

Comment:
CPU/CUDA differ for `unique_dim` in how they handle `return_inverse` and `return_counts`. CPU ignores the arguments and always returns each, whereas CUDA does not. 

I'm not sure if that distinction is important at this level of abstraction, but if it is, I assume we should favor the CUDA implementation?

Reply:
Hmm, this sounds like an eager mode bug. But we have the ability to query the device using `fake_device` so you can give the exact behavior that eager has (and yes, would be good to do that now)

Reply:
To clarify, this difference arises from their source implementations: [CPU](https://github.com/pytorch/pytorch/blob/d62b025efc732a7733850ebfcd097eaf28c8778a/aten/src/ATen/native/Unique.cpp#L358) / [CUDA](https://github.com/pytorch/pytorch/blob/ac519206565e939f1dc9ca632d734caa13a90b85/aten/src/ATen/native/cuda/Unique.cu#L89). One can note that the `return_inverse` and `return_counts` arguments are unused for `_unique_dim_cpu_template`. 

Perhaps this is still something to do with eager mode; I don't know enough about the inner workings and dispatches to rule that out myself.